### ISSUE TITLE
feat(training): build grouped goal-belief datasets, manifests, and a physics-residual trainer

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#8079]
## Current issue-planning review — 2026-09-15

**Audit disposition:** `blocked`. The current audit classifies the issue as blocked; its prerequisite, evidence, or triage gate is incomplete.

**First action:** Inventory current trainer, feature, registry and manifest owners and reconcile the live lifecycle conflict before assigning work.

### Ordered implementation / investigation plan

1. Inventory current trainer, feature, registry and manifest owners and reconcile the live lifecycle conflict before assigning work.
2. Freeze the actor sample schema with causal history ending at t; extract actor features before attaching separately stored oracle labels. Exclude future values, simulator identity, privileged routes and test-fitted priors from inputs.
3. Build deterministic grouped train/validation/held-out/stress partitions using episode, actor-track, route-equivalent and overlapping-window identities. Fit normalizers and candidate priors on train only.
4. Implement H={1,2,3,5,8,16} masks, reset-boundary truncation and effective-history accounting; use oracle-randomization and future-truncation canaries to prove feature independence.
5. Write checksum-bound shards, split and extractor manifests and an exclusion/coverage quality report. Reject corrupt, extra, mixed-schema or missing shards; external assets must pass the existing real-trajectory preflight.
6. Wire tiny mean/linear, MLP and GRU smoke paths to the existing checkpoint contract. Prove learnable-task loss reduction and reload equivalence; graph-GRU may be deferred with a frozen graph schema. Create a separate successor for full training, not a hidden sweep here.

### Files and owners to inspect

- robot_sf/planner/learned_short_horizon_trainer.py
- robot_sf/data_ingestion/real_trajectory_contract.py
- robot_sf/cli_datasets.py
- canonical model registry and training manifest owners
- tests/data_ingestion/ and tests/training/ (proposed focused tests)

### Acceptance and validation

- No episode, overlapping history, track or declared route-equivalent group crosses partitions.
- All losses and data exclusions retain explicit denominators, including censored and unrepresented labels.
- Tiny deterministic CPU smoke or explicitly justified numerical tolerance passes; provenance and checkpoint shape mismatches fail closed.
- Large data remain outside Git; pipeline integrity is not described as learned-model quality.

Named validation commands:

- `uv run python -m robot_sf.training.goal_belief_dataset build \`
- `uv run python -m robot_sf.training.goal_belief_dataset validate \`
- `uv run python -m robot_sf.training.goal_belief_train \`
- `uv run python -m robot_sf.training.goal_belief_checkpoint validate \`
- `scripts/dev/run_worktree_shared_venv.sh -- uv run pytest \`
- `scripts/dev/run_worktree_shared_venv.sh -- uv run ruff check \`
- `scripts/dev/run_worktree_shared_venv.sh -- uv run ruff format --check \`
- `BASE_REF=origin/main scripts/dev/pr_ready_check.sh`

### Current dependencies and issue-text corrections

- Dependencies: #8060 parent; #8063 #8065 #8066 #8075 #8076 #8077; coordinate #8078
- Add a current owner/predecessor and missing-deliverable table; resolve stale state-conflict metadata through the lifecycle owner.
- Make the boundary explicit: dataset/trainer infrastructure and tiny smoke only; full training requires a successor issue and its gates.
- Correct the repeated -k validation example so every intended test group is actually selected.
- The supplied draft was checked against the refreshed live issue and complete comment thread; historical rationale is retained below.
- Keep current out-of-scope, ownership, provenance, rights, compute, and claim boundaries intact. Do not close the issue merely because this plan was posted.

### Lifecycle and claim boundary

- Current state labels: `state:blocked`.
- The issue remains open. No closure, compute submission, release/publication, evidence deletion, or scientific-claim promotion is authorized by this review.
- This review does not request a lifecycle transition; existing assignees, ownership, valid priority labels, and unrelated labels are preserved.
- Proposed non-lifecycle label additions applied with this review: `none`.

**Evidence note:** Full live issue body and complete available comment thread were reviewed during the 2026-09-15 refresh. This is planning metadata, not implementation or benchmark evidence; historical measurements remain historical.

## Archetype Metadata

```yaml
archetype: implementation
evidence_tier: preflight_valid
linked_policy:
  - docs/maintainer_values.md
  - docs/context/issue_1512_issue_archetypes.md
  - docs/context/pedestrian_goal_force_prediction_research.md
```

## Parent and dependencies

- Parent programme: #8060
- Hard predecessors: #8063, #8065, #8066, #8075, #8076, #8077
- Forecast feature compatibility should coordinate with: #8078
- Research/design basis: PR #8058
- Unblocks: learned-model training/selection and later online-calibration work

## Goal

Create a reproducible data and training infrastructure for hybrid pedestrian goal-force prediction that:

- generates causal actor features from observation-only tracks and rule-based beliefs;
- attaches oracle force, waypoint, route, arrival, and change labels only after actor features are frozen;
- preserves episode/track/segment identity and transition timing;
- groups train/validation/test partitions so adjacent frames and related routes cannot leak across splits;
- supports random causal history truncation over `H={1,2,3,5,8,16}`;
- fits normalizers and priors on training data only;
- writes checksum-bound manifests and compact data-quality reports;
- provides a canonical, deterministic trainer interface for linear, MLP, GRU, and later graph models;
- does not commit large raw datasets or claim learned-model quality.

This issue is infrastructure/preflight. The actual training/model-selection campaign belongs to a successor issue.

## Current repository surfaces to reuse

Inspect and extend canonical owners where possible:

- `robot_sf/planner/learned_short_horizon_trainer.py`;
- `robot_sf/planner/learned_short_horizon_predictor.py`;
- `robot_sf/planner/learned_gmm_predictor.py`;
- `robot_sf/data_ingestion/real_trajectory_contract.py` and manifest preflight;
- model registry/checkpoint metadata;
- benchmark episode and transition trace loaders;
- existing dataset CLI surfaces in `robot_sf/cli_datasets.py`;
- any current training manifest/provenance contract.

Do not create a second incompatible model registry, normalizer format, or external-data staging policy.

## Dataset unit and timing

The canonical sample is one tracked pedestrian at one actor decision time, with a causal history ending at time `t` and oracle labels for the transition/event being predicted.

Required identity:

```text
dataset_schema_version
source_repository_commit
source_trace_digest
map_id
scenario_family
scenario_id
episode_id
seed
simulator_pedestrian_id       oracle metadata only
actor_track_id
track_segment_id
goal_segment_id
sample_time_t
history_start_time
history_end_time = t
history_H
future_label_horizon
```

Actor feature extraction must complete before oracle columns are joined. A feature-leakage audit should operate on field names, dependency graph, and randomized-oracle canaries.

### Required actor features

Per history step, where available:

- tracked position and velocity in canonical frame;
- position/velocity covariance and association confidence;
- causal acceleration mean/covariance;
- actor-reconstructed typed force components and availability masks;
- residual goal-force estimate and covariance;
- robot state/action/plan visible to the actor;
- local neighbour state or graph edges from visible tracks;
- obstacle/local map descriptors available to the planner;
- candidate IDs represented through stable embeddings/features, not raw test-only labels;
- candidate direction/path tangent, distance, role, feasibility, and public prior;
- rule-based waypoint/route/unknown probabilities;
- arrival/change/mode diagnostics;
- saturation, visibility, lost-track, partial-force, and validity masks;
- actual `dt` and history age.

### Prohibited actor features

- true active/final goal coordinates unless they are also public candidates and only public candidate geometry is used;
- true assigned route or waypoint index;
- oracle force components in an actor-trained arm;
- simulator pedestrian identity as a predictive feature;
- future positions/velocities/events;
- labels or normalizers fitted using validation/test data;
- issue/scenario identifiers that act as target leakage unless explicitly ablated and justified.

### Optional graph representation

If graph training is enabled, define:

- target pedestrian node;
- neighbour nodes with actor-visible features;
- robot node or global context;
- edge features from relative state and interaction geometry;
- masks and deterministic node ordering;
- maximum node count and truncation policy;
- no oracle group membership unless publicly observable and a dedicated arm declares it.

## Labels

Version and store labels separately from actor features:

### Continuous labels

- true desired goal-force vector;
- true desired direction;
- preferred speed and force parameters for auxiliary/upper-bound tasks;
- next velocity/position residual;
- future positions for forecast-aware auxiliary training.

### Categorical/probabilistic labels

- true active candidate when represented;
- candidate-not-represented/unknown;
- true final destination/route signature when representable;
- motion/event class;
- change/no-change and event subtype;
- arrival/stopped/restart;
- speed-censored state;
- component-observability state;
- tracking association correctness for stratification, not actor input.

Labels must respect #8063 transition timing and #8076 event definitions.

## History truncation

Train one model family across multiple causal history lengths:

- sample `H` from the declared set using a versioned distribution;
- mask unavailable prefix rows;
- never synthesize `H=1` acceleration features from older frames;
- after a reset/change boundary, prohibit histories from crossing the boundary unless the feature is explicitly route-level and contractually permitted;
- record effective valid history, not only requested `H`;
- test that changing future frames cannot change current features.

## Dataset partitions

Partition by grouped identities, not samples. Required split keys:

- map;
- scenario family;
- route family or topology signature;
- episode seed;
- goal-change pattern;
- optionally pedestrian-model/parameter regime.

At minimum define:

1. **development/train:** fitting weights/normalizers;
2. **validation:** hyperparameters, thresholds, calibration, early stopping;
3. **held-out confirmation:** no tuning;
4. **stress/transfer:** unseen maps/routes, parameter shifts, occlusion, missing candidates.

No episode, actor track, route-equivalent sequence, or adjacent window may cross partitions. If a route occurs on multiple maps, decide whether route topology is a grouping key and document the transfer question.

## Leakage audits

Mandatory automated checks:

1. no shared `episode_id` across partitions;
2. no shared source transition/track segment across partitions;
3. no overlapping temporal windows across partitions;
4. normalizer fit record references train partition only;
5. candidate-prior extraction references train partition only;
6. oracle field names absent from actor feature schema;
7. randomizing oracle labels leaves feature bytes identical;
8. future trajectory truncation leaves current features identical;
9. partition generation deterministic under input row permutation;
10. duplicate source rows detected and rejected;
11. source trace and manifest digest verified before loading;
12. external data cannot be used unless its manifest passes the repository's real-trajectory preflight.

## Data-quality report

Produce a compact report containing:

- sample/episode/track/segment counts per partition;
- maps, scenarios, routes, force models, and event types;
- `H` distribution and valid-history coverage;
- candidate coverage and unknown rates;
- class prevalence for change/arrival/event labels;
- saturation, partial, occluded, lost, unavailable, and fallback rates;
- feature missingness and finite-value checks;
- force/direction label distributions;
- tracking correctness strata;
- duplicate/leakage audit result;
- split balance and held-out uniqueness;
- source/config/schema digests.

Do not silently drop invalid rows. Record every exclusion reason and denominator.

## Serialization and manifests

Use a format appropriate for large numeric data, such as Parquet/NPZ shards, subject to current repository conventions. Requirements:

- deterministic shard ordering;
- schema version and dtypes/shapes;
- per-file SHA-256;
- row count and identity range;
- source trace digest;
- feature extractor/config hash;
- split assignment digest;
- license/provenance for external data;
- normalizer digest;
- no raw large data committed to Git;
- compact manifest and quality report may be committed after review.

The loader must fail closed on missing, extra, corrupt, mixed-version, or hash-mismatched shards.

## Trainer interface

Provide a canonical trainer protocol/config supporting at least:

```text
model_type = ridge | mlp | gru | graph_gru
seed
device
history_lengths
feature_schema_version
output_heads
loss weights
optimizer/learning rate
batch size
epochs/early stopping
train/validation manifest
normalizer
checkpoint/output directory
```

The infrastructure issue need only prove tiny deterministic training for simple baselines and one recurrent path. Actual full sweeps belong to the successor campaign.

Required outputs:

- checkpoint with state dict and architecture metadata;
- training config and hashes;
- feature/label schema versions;
- normalizer;
- train/validation metric history;
- RNG/environment metadata;
- model card/claim boundary;
- registry-compatible manifest;
- exact command.

## Baseline models in infrastructure smoke

Implement or wire tiny deterministic smoke support for:

1. zero/mean predictor;
2. ridge or linear residual regressor;
3. small MLP;
4. small GRU with history mask.

A graph-GRU implementation may be deferred to the model-selection issue if the graph feature contract is frozen here. The smoke proves the training pipeline, not model quality.

## Loss interface

Support named heads/losses, including:

- Gaussian or mixture NLL for goal force;
- cosine/directional loss;
- candidate/route cross entropy with unknown class;
- change/arrival binary or multiclass loss;
- one-step dynamics residual loss;
- calibration regularizer where explicitly defined;
- masks for censored/unrepresented/unavailable labels.

Every loss denominator must be reported. Censored force magnitude and absent true candidate cannot enter incompatible heads.

## Determinism

For CPU smoke with fixed seed/thread environment:

- split manifest digest identical across reruns;
- feature shard digest identical across reruns;
- tiny training batch order identical;
- initial model state digest identical;
- loss history and checkpoint digest identical where deterministic backend support permits;
- otherwise define a strict numerical tolerance and record nondeterministic backend state.

No “best effort” determinism claim without environment metadata.

## Required tests

Suggested files:

- `tests/data_ingestion/test_goal_belief_dataset_manifest.py`
- `tests/training/test_goal_belief_feature_extractor.py`
- `tests/training/test_goal_belief_split_integrity.py`
- `tests/training/test_goal_belief_normalizer.py`
- `tests/training/test_goal_belief_trainer.py`
- `tests/training/test_goal_belief_checkpoint_contract.py`

Required cases:

- exact feature/label shapes and masks for every `H`;
- no future/oracle leakage canaries;
- reset boundary truncation;
- deterministic grouped split and no overlap;
- duplicate/corrupt/mixed-schema shard rejection;
- train-only normalizer fit;
- candidate-not-represented masking;
- censored force loss masking;
- class-imbalance/prevalence report;
- tiny ridge/MLP/GRU loss decreases on a learnable synthetic task;
- checkpoint round-trip produces identical predictions;
- checkpoint feature schema mismatch fails closed;
- missing torch/external dependency has a clear remedy;
- external manifest unavailable/invalid remains blocked, not synthetic nominal evidence.

## Config-first pipeline

Provide canonical commands, for example:

```bash
uv run python -m robot_sf.training.goal_belief_dataset build \
  --config configs/training/goal_belief_dataset_smoke.yaml \
  --output output/training/goal_belief_dataset_smoke

uv run python -m robot_sf.training.goal_belief_dataset validate \
  --manifest output/training/goal_belief_dataset_smoke/manifest.json

uv run python -m robot_sf.training.goal_belief_train \
  --config configs/training/goal_belief_trainer_smoke.yaml \
  --output output/models/goal_belief_trainer_smoke

uv run python -m robot_sf.training.goal_belief_checkpoint validate \
  --manifest output/models/goal_belief_trainer_smoke/model_manifest.json
```

Use current CLI/owner conventions if different.

## Canonical validation procedure

The PR must report exact results for commands equivalent to:

```bash
scripts/dev/run_worktree_shared_venv.sh -- uv run pytest \
  tests/data_ingestion -k "goal_belief or manifest" \
  tests/training -k "goal_belief" \
  tests/models -k "checkpoint or registry" -q

scripts/dev/run_worktree_shared_venv.sh -- uv run ruff check \
  robot_sf/data_ingestion robot_sf/training robot_sf/models \
  tests/data_ingestion tests/training tests/models

scripts/dev/run_worktree_shared_venv.sh -- uv run ruff format --check \
  robot_sf/data_ingestion robot_sf/training robot_sf/models \
  tests/data_ingestion tests/training tests/models

BASE_REF=origin/main scripts/dev/pr_ready_check.sh
```

Adjust paths to canonical owners discovered during implementation.

## Acceptance criteria

- [ ] Actor features are generated before and independently of oracle labels.
- [ ] Feature schema excludes oracle goal/route/identity/future fields.
- [ ] Grouped train/validation/test/stress splits have no episode/window/track leakage.
- [ ] `H={1,2,3,5,8,16}` masks and reset boundaries are correct.
- [ ] Candidate-missing, censored, partial, lost, and unavailable labels are explicit.
- [ ] Train-only normalizers/priors and their digests are enforced.
- [ ] Manifests verify every shard and source trace.
- [ ] Tiny baseline/recurrent training and checkpoint round-trip work deterministically.
- [ ] Model registry/feature-contract checks fail closed.
- [ ] External data remains blocked unless validated through existing provenance policy.
- [ ] Evidence is pipeline/preflight integrity, not learned-model performance.

## Falsification and stop rule

Do not launch a learned-model campaign if any split leakage, oracle feature dependency, future-data dependency, normalizer contamination, mixed schema, or non-reproducible source manifest remains. Mark the dataset blocked and repair the producer rather than compensating in training code.

## Risks and rollback

- Huge data volume: shard, stream, and commit only manifests/quality summaries.
- Severe class imbalance: report prevalence before weighting; preserve natural evaluation distribution.
- Feature contract churn: version schemas and make checkpoints fail closed.
- Synthetic overfitting: keep synthetic smoke clearly separate from real/benchmark evidence.
- Rollback: remove dataset/trainer adapters; rule-based inference remains available.

## Evidence output and downstream propagation

Commit compact preflight manifests/quality receipts and update #8060. Do not register a model as benchmark-eligible or claim prediction improvement until the successor training/validation issue passes Gate E.

```yaml
schema: goal_autopilot_preparation.v1
repository: ll7/robot_sf_ll7
issue: 8079
source_body_sha256: 6aeab7ff44046acd0e07510c7e256086c9e8eadd7238083bfcf7e407c8b325c5
source_comments_sha256: 
audit_schema: open_issue_contract_audit.v1
audit_digest: 1c191411793929640684e0827d763f3a2009a8682ab98dcefa8c2ded3e54f4ea
audit_classification: state_conflict
next_action: reconcile_state_labels
authority: lifecycle_owner
execution_mode: lifecycle
preferred_worker: MaxRunner
expected_pr_runner_label: runner:max
implementation_admitted: False
state_ready_change_proposed: False
mutation_batch: open-issues-20260830-06
```

This packet is preparation evidence only. It never overrides live labels, exact claim state, branch state, typed dependencies, domain gates, compute authority, release authority, or scientific evidence rules.
